File: src/templates/tmpl.ml
(* OCaml *)

let b j=Char.code "<QCImage/>\000".[j]
(* ---- state (a=acc, i=instruction pos, p=READ pos, r=registers + () return stack, s=stack top, d=skip depth) ---- *)
let a=ref 0 let i=ref <QCConst>entrypoint</QCConst> let p=ref 0 let r=Array.make 99 0 let s=ref 8 let d=ref 0

(* ---- u = read and advance; then put the name into r[0..] ---- *)
let u()=(i:= !i+1;b(!i-1))let()=String.iteri(fun c x->Array.set r c(Char.code x))(Array.append Sys.argv[|"ml"|]).(1);

(* ---- main loop (an else-if chain over ranges of c) ---- *)
while !i<<QCConst>image_len</QCConst> do let c=u()in if c<34 then((while b(!i)<>47 do(let c=u()in print_char(Char.chr(if c=72 then u()-33 else c)))done);
  i:= !i+1)else if c<39 then(a:=b(!p);
  p:= !p+1)else if c<41 then(Array.set r(!s)(!i);s:= !s+1;
  d:=1;
  while !d>0 do(i:= !i+1;
    let c=b(!i)in if c/2=20 then d:= !d+81-2*c)done)
else if c<42 then(if !a=0 then s:= !s-1 else i:=Array.get r(!s-1))
else if c<43 then print_char(Char.chr(!a))else if c<56 then Array.set r(c mod 8)(!a)
else if c<64 then a:=Array.get r(c mod 8)else if c<72 then a:= !a-Array.get r(c mod 8)else a:=c done

File: src/templates/tmpl.fs
\ Forth

: d dup <QCConst>image_len</QCConst> < if s" <QCImage/>" drop + c@ else drop 0 then ;

\ ---- state ----
0 value a 0 value b 0 value p 0 value l 0 value r 0 value c
create y 99 cells allot

\ ---- main: init the state, put the name into r[0..], then run (names are at most 6 chars, so no mod 8) ----
: s y 8 cells erase 8 to l <QCConst>entrypoint</QCConst> to b
  next-arg dup 0= if 2drop s" fs" then
  0 ?do dup I + c@ I cells y + ! loop drop
  begin b d b 1 + to b dup while to c
    c 34 < if begin b d b 1 + to b dup 47 <> while dup 72 = if drop b d b 1 + to b 33 - then emit repeat drop
    else c 39 < if p d to a p 1 + to p
    else c 41 < if b l cells y + ! l 1 + to l 1 to r
      begin r while b 1 + to b b d 2/ 20 = if r 81 b d 2* - + to r then repeat
    else c 42 < if a if l 1- cells y + @ to b else l 1- to l then
    else c 43 < if a emit
    else c 56 < if a c 8 mod cells y + !
    else c 64 < if c 8 mod cells y + @ else c 72 < if a c 8 mod cells y + @ - else c then then to a
    then then then then then then
  repeat drop ;
s bye
